File: FBSimulatorControl/Strategies/FBDefaultsModificationStrategy.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDefaultsModificationStrategy.h"

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimRuntime.h>

#import "FBSimulator.h"
#import "FBSimulatorError.h"
#import "FBSimulatorLaunchCtlCommands.h"

@interface FBDefaultsModificationStrategy ()

@property (nonatomic, strong, readonly) FBSimulator *simulator;

@end

@implementation FBDefaultsModificationStrategy

+ (instancetype)strategyWithSimulator:(FBSimulator *)simulator
{
  return [[self alloc] initWithSimulator:simulator];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;
  return self;
}

- (NSString *)defaultsBinary
{
  NSString *path = [[[self.simulator.device.runtime.root
    stringByAppendingPathComponent:@"usr"]
    stringByAppendingPathComponent:@"bin"]
    stringByAppendingPathComponent:@"defaults"];
  NSError *error = nil;
  FBBinaryDescriptor *binary = [FBBinaryDescriptor binaryWithPath:path error:&error];
  NSAssert(binary, @"Could not locate defaults at expected location '%@', error %@", path, error);
  return binary.path;
}

- (FBFuture<NSNull *> *)modifyDefaultsInDomainOrPath:(NSString *)domainOrPath defaults:(NSDictionary<NSString *, id> *)defaults
{
  NSError *innerError = nil;
  NSString *file = [self.simulator.auxillaryDirectory stringByAppendingPathComponent:@"temporary.plist"];
  if (![NSFileManager.defaultManager createDirectoryAtPath:[file stringByDeletingLastPathComponent] withIntermediateDirectories:YES attributes:nil error:&innerError]) {
    return [[[FBSimulatorError
      describeFormat:@"Could not create intermediate directories for temporary plist %@", file]
      causedBy:innerError]
      failFuture];
  }
  if (![defaults writeToFile:file atomically:YES]) {
    return [[FBSimulatorError
      describeFormat:@"Failed to write out defaults to temporary file %@", file]
      failFuture];
  }

  // Build the arguments
  NSMutableArray<NSString *> *arguments = [NSMutableArray arrayWithObject:@"import"];
  if (domainOrPath) {
    [arguments addObject:domainOrPath];
  }
  [arguments addObject:file];

  return [[self performDefaultsCommandWithArguments:arguments] mapReplace:NSNull.null];
}

- (FBFuture<NSNull *> *)setDefaultInDomain:(NSString *)domain key:(NSString *)key value:(NSString *)value type:(NSString *)type
{
  return [[self
    performDefaultsCommandWithArguments:@[
      @"write",
      domain,
      key,
      [NSString stringWithFormat:@"-%@", type ? type : @"string"],
      value,
    ]]
    mapReplace:NSNull.null];
}

- (FBFuture<NSString *> *)getDefaultInDomain:(NSString *)domain key:(NSString *)key
{
  return [self
    performDefaultsCommandWithArguments:@[
      @"read",
      domain,
      key,
    ]];
}

- (FBFuture<NSString *> *)performDefaultsCommandWithArguments:(NSArray<NSString *> *)arguments
{
  // Make the Launch Config
  FBProcessSpawnConfiguration *configuration = [[FBProcessSpawnConfiguration alloc]
    initWithLaunchPath:self.defaultsBinary
    arguments:arguments
    environment:@{}
    io:FBProcessIO.outputToDevNull
    mode:FBProcessSpawnModeDefault];

  // Run the defaults command.
  return [[FBProcessSpawnCommandHelpers
    launchConsumingStdout:configuration withCommands:self.simulator]
    onQueue:self.simulator.asyncQueue map:^(NSString *output) {
      return [output stringByTrimmingCharactersInSet:NSCharacterSet.newlineCharacterSet];
    }];
}

- (FBFuture<NSNull *> *)amendRelativeToPath:(NSString *)relativePath defaults:(NSDictionary<NSString *, id> *)defaults managingService:(NSString *)serviceName
{
  FBSimulator *simulator = self.simulator;
  FBiOSTargetState state = simulator.state;
  if (state != FBiOSTargetStateBooted && state != FBiOSTargetStateShutdown) {
    return [[FBSimulatorError
      describeFormat:@"Cannot amend a plist when the Simulator state is %@, should be %@ or %@", FBiOSTargetStateStringFromState(state), FBiOSTargetStateStringShutdown, FBiOSTargetStateStringBooted]
      failFuture];
  }

  // Stop the service, if booted.
  FBFuture<NSNull *> *stopFuture = state == FBiOSTargetStateBooted
    ? [[simulator stopServiceWithName:serviceName] mapReplace:NSNull.null]
    : FBFuture.empty;

  // The path to amend.
  NSString *fullPath = [self.simulator.dataDirectory stringByAppendingPathComponent:relativePath];

  return  [[stopFuture
    onQueue:self.simulator.workQueue fmap:^FBFuture *(NSNull *_) {
      return [self modifyDefaultsInDomainOrPath:fullPath defaults:defaults];
    }]
    onQueue:self.simulator.workQueue fmap:^FBFuture<NSNull *> *(NSNull *_) {
      // Re-start the Service if booted.
      return state == FBiOSTargetStateBooted
        ? [[simulator startServiceWithName:serviceName] mapReplace:NSNull.null]
        : FBFuture.empty;
    }];
}

@end

@implementation FBPreferenceModificationStrategy

static NSString *const AppleGlobalDomain = @"Apple Global Domain";

- (FBFuture<NSNull *> *)setPreference:(NSString *)name value:(NSString *)value type:(nullable NSString *)type domain:(nullable NSString *)domain
{
  if (domain == nil) {
    domain = AppleGlobalDomain;
  }
  return [self setDefaultInDomain:domain key:name value:value type:type];
}

- (FBFuture<NSString *> *)getCurrentPreference:(NSString *)name domain:(nullable NSString *)domain
{
  if (domain == nil) {
    domain = AppleGlobalDomain;
  }
  return [self getDefaultInDomain:domain key:name];
}

@end

@implementation FBLocationServicesModificationStrategy

- (FBFuture<NSNull *> *)approveLocationServicesForBundleIDs:(NSArray<NSString *> *)bundleIDs
{
  NSParameterAssert(bundleIDs);

  NSMutableDictionary<NSString *, id> *defaults = [NSMutableDictionary dictionary];
  for (NSString *bundleID in bundleIDs) {
    defaults[bundleID] = @{
      @"Whitelisted": @NO,
      @"BundleId": bundleID,
      @"SupportedAuthorizationMask" : @3,
      @"Authorization" : @2,
      @"Authorized": @YES,
      @"Executable": @"",
      @"Registered": @"",
    };
  }

  return [self
    amendRelativeToPath:@"Library/Caches/locationd/clients.plist"
    defaults:[defaults copy]
    managingService:@"locationd"];
}

- (FBFuture<NSNull *> *)revokeLocationServicesForBundleIDs:(NSArray<NSString *> *)bundleIDs
{
  NSParameterAssert(bundleIDs);

    FBSimulator *simulator = self.simulator;
    FBiOSTargetState state = simulator.state;
    if (state != FBiOSTargetStateBooted && state != FBiOSTargetStateShutdown) {
      return [[FBSimulatorError
        describeFormat:@"Cannot modify a plist when the Simulator state is %@, should be %@ or %@", FBiOSTargetStateStringFromState(state), FBiOSTargetStateStringShutdown, FBiOSTargetStateStringBooted]
        failFuture];
    }

    NSString *serviceName = @"locationd";

    // Stop the service, if booted.
    FBFuture<NSNull *> *stopFuture = state == FBiOSTargetStateBooted
      ? [[simulator stopServiceWithName:serviceName] mapReplace:NSNull.null]
      : FBFuture.empty;

    NSString *path = [self.simulator.dataDirectory
                      stringByAppendingPathComponent:@"Library/Caches/locationd/clients.plist"];
    NSMutableArray<FBFuture<NSString *> *> *futures = [NSMutableArray array];
    for (NSString *bundleID in bundleIDs) {
      [futures addObject:
       [self
        performDefaultsCommandWithArguments:@[
          @"delete",
          path,
          bundleID,
        ]]];
    }

    return [[stopFuture
      onQueue:self.simulator.workQueue fmap:^FBFuture *(NSNull *_) {
        return [[FBFuture futureWithFutures:futures] mapReplace:NSNull.null];
      }]
      onQueue:self.simulator.workQueue fmap:^FBFuture<NSNull *> *(NSNull *_) {
        // Re-start the Service if booted.
        return state == FBiOSTargetStateBooted
          ? [[simulator startServiceWithName:serviceName] mapReplace:NSNull.null]
          : FBFuture.empty;
      }];
}

@end

File: FBSimulatorControl/Strategies/FBDefaultsModificationStrategy.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
import Foundation

// swiftlint:disable force_unwrapping

public class FBDefaultsModificationStrategy: NSObject {

  // MARK: - Properties

  fileprivate let simulator: FBSimulator

  // MARK: - Initializers

  required init(simulator: FBSimulator) {
    self.simulator = simulator
    super.init()
  }

  // MARK: - Public Methods

  public func modifyDefaults(inDomainOrPath domainOrPath: String?, defaults: [String: Any]) async throws {
    let file = (simulator.auxillaryDirectory as NSString).appendingPathComponent("temporary.plist")
    let dirPath = (file as NSString).deletingLastPathComponent

    do {
      try FileManager.default.createDirectory(atPath: dirPath, withIntermediateDirectories: true, attributes: nil)
    } catch {
      throw
        FBSimulatorError
        .describe("Could not create intermediate directories for temporary plist \(file)")
        .caused(by: error as NSError)
        .build()
    }

    if !(defaults as NSDictionary).write(toFile: file, atomically: true) {
      throw FBSimulatorError.describe("Failed to write out defaults to temporary file \(file)").build()
    }

    _ = try await run(.importPlist(domainOrPath: domainOrPath, file: file))
  }

  // MARK: - Internal Methods

  fileprivate func setDefault(inDomain domain: String, key: String, value: String, type: String?) async throws {
    _ = try await run(.write(domain: domain, key: key, type: type ?? "string", value: value))
  }

  fileprivate func getDefault(inDomain domain: String, key: String) async throws -> NSString {
    return try await run(.read(domain: domain, key: key))
  }

  // The closed set of `defaults` operations this strategy issues. Modelling them as an enum keeps argv
  // construction in one place and makes the operation set exhaustive, so a new operation cannot be
  // added without routing through `run`.
  enum Command {
    case read(domain: String, key: String)
    case write(domain: String, key: String, type: String, value: String)
    case importPlist(domainOrPath: String?, file: String)
    case delete(path: String, key: String)

    var arguments: [String] {
      switch self {
      case let .read(domain, key):
        return ["read", domain, key]
      case let .write(domain, key, type, value):
        return ["write", domain, key, "-\(type)", value]
      case let .importPlist(domainOrPath, file):
        var args = ["import"]
        if let domainOrPath {
          args.append(domainOrPath)
        }
        args.append(file)
        return args
      case let .delete(path, key):
        return ["delete", path, key]
      }
    }

    var exitCodePolicy: ExitCodePolicy {
      switch self {
      case .read, .delete:
        // `defaults` returns 1 for a missing key/domain (a benign optional read or idempotent delete)
        // and for a genuine failure alike, with no distinguishing code, so tolerate any non-zero.
        return .tolerateAny
      case .write, .importPlist:
        return .require([0])
      }
    }
  }

  fileprivate func run(_ command: Command) async throws -> NSString {
    let launchPath = defaultsBinary
    let output = try await simulator.launchProcessConsumingOutput(launchPath: launchPath, arguments: command.arguments)
    return try FBDefaultsModificationStrategy.stdout(orThrowFrom: output, command: command, logger: simulator.logger)
  }

  // Internal for unit-test coverage of the exit-code handling; see FBDefaultsModificationStrategyTests.
  static func stdout(orThrowFrom output: FBInSimulatorToolOutput, command: Command, logger: (any FBControlCoreLogger)?) throws -> NSString {
    if output.exitCode != 0 {
      let stderr = String(data: output.stderr, encoding: .utf8) ?? ""
      guard command.exitCodePolicy.accepts(output.exitCode) else {
        throw FBSimulatorError.describe("defaults \(command.arguments.joined(separator: " ")) failed with exit code \(output.exitCode): \(stderr)").build()
      }
      logger?.log("defaults \(command.arguments.joined(separator: " ")) exited with code \(output.exitCode): \(stderr)")
    }
    let stdout = String(data: output.stdout, encoding: .utf8) ?? ""
    return stdout.trimmingCharacters(in: .newlines) as NSString
  }

  fileprivate func amendRelativeTo(path relativePath: String, defaults: [String: Any], managingService serviceName: String) async throws {
    let state = simulator.state
    guard state == .booted || state == .shutdown else {
      throw
        FBSimulatorError
        .describe("Cannot amend a plist when the Simulator state is \(FBiOSTargetStateStringFromState(state)), should be \(FBiOSTargetStateString.shutdown) or \(FBiOSTargetStateString.booted)")
        .build()
    }

    // Stop the service while the plist is rewritten, restarting it afterwards if it was running.
    if state == .booted {
      _ = try await simulator.stopService(withName: serviceName)
    }
    let fullPath = (simulator.dataDirectory! as NSString).appendingPathComponent(relativePath)
    try await modifyDefaults(inDomainOrPath: fullPath, defaults: defaults)
    if state == .booted {
      _ = try await simulator.startService(withName: serviceName)
    }
  }

  // MARK: - Private

  private var defaultsBinary: String {
    let path =
      ((simulator.device.runtime.root! as NSString)
      .appendingPathComponent("usr") as NSString)
      .appendingPathComponent("bin") as NSString
    let fullPath = path.appendingPathComponent("defaults")
    do {
      let binary = try FBBinaryDescriptor.binary(withPath: fullPath)
      return binary.path
    } catch {
      fatalError("Could not locate defaults at expected location '\(fullPath)', error \(error)")
    }
  }
}

// MARK: - FBPreferenceModificationStrategy

public class FBPreferenceModificationStrategy: FBDefaultsModificationStrategy {

  private static let appleGlobalDomain = "Apple Global Domain"

  public func setPreference(_ name: String, value: String, type: String?, domain: String?) async throws {
    let effectiveDomain = domain ?? FBPreferenceModificationStrategy.appleGlobalDomain
    try await setDefault(inDomain: effectiveDomain, key: name, value: value, type: type)
  }

  public func getCurrentPreference(_ name: String, domain: String?) async throws -> String {
    let effectiveDomain = domain ?? FBPreferenceModificationStrategy.appleGlobalDomain
    return try await getDefault(inDomain: effectiveDomain, key: name) as String
  }
}

// MARK: - FBLocationServicesModificationStrategy

public class FBLocationServicesModificationStrategy: FBDefaultsModificationStrategy {

  public func approveLocationServices(forBundleIDs bundleIDs: [String]) async throws {
    var defaults: [String: Any] = [:]
    for bundleID in bundleIDs {
      defaults[bundleID] =
        [
          "Whitelisted": false,
          "BundleId": bundleID,
          "SupportedAuthorizationMask": 3,
          "Authorization": 2,
          "Authorized": true,
          "Executable": "",
          "Registered": "",
        ] as [String: Any]
    }

    try await amendRelativeTo(
      path: "Library/Caches/locationd/clients.plist",
      defaults: defaults,
      managingService: "locationd"
    )
  }

  public func revokeLocationServices(forBundleIDs bundleIDs: [String]) async throws {
    let state = simulator.state
    guard state == .booted || state == .shutdown else {
      throw
        FBSimulatorError
        .describe("Cannot modify a plist when the Simulator state is \(FBiOSTargetStateStringFromState(state)), should be \(FBiOSTargetStateString.shutdown) or \(FBiOSTargetStateString.booted)")
        .build()
    }

    let serviceName = "locationd"
    if state == .booted {
      _ = try await simulator.stopService(withName: serviceName)
    }

    let path = (simulator.dataDirectory! as NSString)
      .appendingPathComponent("Library/Caches/locationd/clients.plist")
    // Delete sequentially: every delete is a read-modify-write of the same clients.plist, so running
    // them concurrently races and can drop entries when revoking several bundle IDs at once.
    for bundleID in bundleIDs {
      _ = try await run(.delete(path: path, key: bundleID))
    }

    if state == .booted {
      _ = try await simulator.startService(withName: serviceName)
    }
  }
}
